sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +306,132 @@ case class GpuCaseWhen(
   }
 
   override def columnarEval(batch: ColumnarBatch): Any = {
-    // `elseRet` will be closed in `computeIfElse`.
-    val elseRet = elseValue
-      .map(_.columnarEval(batch))
-      .getOrElse(GpuScalar(null, branches.last._2.dataType))
-    branches.foldRight[Any](elseRet) { case ((predicateExpr, trueExpr), falseRet) =>
-      computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+    if (branchesWithSideEffects) {
+      columnarEvalWithSideEffects(batch)
+    } else {
+      // `elseRet` will be closed in `computeIfElse`.
+      val elseRet = elseValue
+        .map(_.columnarEval(batch))
+        .getOrElse(GpuScalar(null, branches.last._2.dataType))
+      branches.foldRight[Any](elseRet) {
+        case ((predicateExpr, trueExpr), falseRet) =>
+          computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+        }
+      }
+    }
+
+  /**
+   * Perform lazy evaluation of each branch sa that we only evaluate the THEN expressions
+   * against rows where the WHEN expression is true.
+   */
+  private def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    // track cumulative state of predicates so that never evaluate expressions
+    // if an earlier expression has already been evaluated
+    var cumulativePred: Option[GpuColumnVector] = None
+
+    // this variable contains the currently evaluated value for each row and gets updated
+    // as each branch is evaluated
+    var currentValue: Option[GpuColumnVector] = None
+
+    try {
+      withResource(GpuColumnVector.from(batch)) { tbl =>
+
+        // iterate over the WHEN THEN branches first
+        branches.foreach {
+          case (whenExpr, thenExpr) =>
+            // evaluate the WHEN predicate
+            withResource(GpuExpressionsUtils.columnarEvalToColumn(whenExpr, batch)) { whenBool =>
+              // we only want to evaluate where this WHEN is true and no previous WHEN has been true
+              val whenBoolAdjusted: GpuColumnVector = cumulativePred match {
+                case Some(prev) =>
+                  withResource(prev.getBase.not()) { notPreviouslyTrue =>
```

Comment:
There's an early-out opportunity where `notPreviouslyTrue` is all false.  In that case all further branches can be avoided since all inputs have been covered by previous branches.  It's tempting to test for that here, but we already evaluated `whenBool` by this point, and that would be wasted computation if we early-out.  It's also tempting to evaluate it after we update `cumulativePred` but that would be wasteful if this is the last branch to evaluate.  So a bit tricky to implement, but I could see it being very beneficial in many case when statements, especially when they are elaborate.

Reply:
I pushed an optimization for this, checking `cumulativePred` after we set it. Even if we are processing the last `WHEN` branch, we can still skip the `ELSE` branch which comes after the main loop.